### PR TITLE
Drop messages in case of no connection

### PR DESCRIPTION
**Pull request by miiu96**
- The implemented mechanism incorporates a feature that allows the skipping of message sending when there is no established connection with a client or a server. When such a connection is absent, the behavior of the system depends on a configurable flag. If the flag is set to false, the messages will be silently discarded without any indication. However, if the flag is set to true, indicating a preference for error notifications, the send method will return an error when attempting to send messages in the absence of a connection. This mechanism offers flexibility and control over how the system handles message delivery in different connectivity scenarios.

### CODE REVIEW

File: websocket/connection/wsConnClient_test.go [hunk 1]
```diff
@@ -161,3 +161,18 @@ func TestWsConnClient_ReOpenAlreadyOpenedConnectionShouldError(t *testing.T) {
 
 	_ = conClient.Close()
 }
+
+func TestWsConnClient_IsOpen(t *testing.T) {
+	testServer := testscommon.NewHttpTestEchoHandler()
+	defer testServer.Close()
```
Comment (mariusmihaic) [inline websocket/connection/wsConnClient_test.go:167]: Doesn't this `Close` return an error and this defer causes a waning on the IDE (smth like unhandled error)?

File: websocket/client/client.go [hunk 2]
```diff
@@ -128,6 +131,11 @@ func (c *client) start() {
 
 // Send will send the provided payload from args
 func (c *client) Send(payload []byte, topic string) error {
+	skipSend := !c.blockingSendIfNoConnection && !c.wsConn.IsOpen()
```
Comment (bogdan-rosianu) [inline websocket/client/client.go:134]: skipSend -> shouldNotSendIfNoConnection

File: websocket/client/client.go [hunk 3]
```diff
@@ -17,21 +17,23 @@ import (
...
-	RetryDurationInSeconds int
-	WithAcknowledge        bool
-	BlockingAckOnError     bool
-	URL                    string
-	PayloadConverter       websocket.PayloadConverter
-	Log                    core.Logger
+	RetryDurationInSeconds     int
+	WithAcknowledge            bool
+	BlockingAckOnError         bool
+	BlockingSendIfNoConnection bool
```
Comment (bogdan-rosianu) [inline websocket/client/client.go:23]: as discussed, maybe rename to `DropMessagesIfNoConnection`

File: websocket/client/client_test.go [hunk 4]
```diff
@@ -104,3 +106,17 @@ func TestClient_Send(t *testing.T) {
 
 	require.Equal(t, uint64(1), atomic.LoadUint64(&count))
 }
+
+func TestClient_ShouldSkipSend(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply websocket/client/client_test.go]: renamed

File: websocket/client/client.go (hunk 2 above)
Comment (miiu96) [reply websocket/client/client.go]: renamed

File: websocket/client/client.go (hunk 3 above)
Comment (miiu96) [reply websocket/client/client.go]: renamed

File: websocket/server/server.go [hunk 5]
```diff
@@ -143,7 +147,17 @@ func (s *server) initializeServer(wsURL string, wsPath string) {
 
 // Send will send the provided payload from args
 func (s *server) Send(payload []byte, topic string) error {
-	for _, tuple := range s.transceiversAndConn.getAll() {
+	transceiversAndCon := s.transceiversAndConn.getAll()
+	noConnection := len(transceiversAndCon) == 0
+	skipSendNoConnection := noConnection && !s.blockingSendIfNoConnection
```
[the remark this answers is not in the document]
Comment (miiu96) [reply websocket/server/server.go]: renamed

File: websocket/server/server_test.go [hunk 6]
```diff
@@ -108,3 +108,31 @@ func TestServer_ListenAndRegisterPayloadHandlerAndClose(t *testing.T) {
...
+	args.BlockingSendIfNoConnection = true
+	args.URL = "localhost:9211"
+	wsServer, _ := NewWebSocketServer(args)
+
+	defer func() {
+		_ = wsServer.Close()
+	}()
+
+	err := wsServer.Send([]byte("test"), "test")
+	require.Equal(t, "cannot send message: no clients connected", err.Error())
```
[the remark this answers is not in the document]
Comment (miiu96) [reply websocket/server/server_test.go:137]: done

File: websocket/data/operations.go [hunk 7]
```diff
@@ -11,9 +11,10 @@ const (
...
-	Mode               string
-	RetryDurationInSec int
-	WithAcknowledge    bool
-	BlockingAckOnError bool
+	URL                        string
+	Mode                       string
+	RetryDurationInSec         int
+	WithAcknowledge            bool
+	BlockingAckOnError         bool
+	BlockingSendIfNoConnection bool
```
[the remark this answers is not in the document]
Comment (miiu96) [reply websocket/data/operations.go]: refactored

File: websocket/client/client_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (miiu96) [reply websocket/client/client_test.go]: renamed

**Review by bogdan-rosianu — APPROVED** [review]

**Review by mariusmihaic — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Pull request merged**